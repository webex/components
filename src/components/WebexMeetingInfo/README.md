# Webex Meeting Info Component

Webex Meeting Info component displays information about a meeting.

<p align="center">
  <img src="./WebexMeetingInfo.png" alt="Default Webex Meeting Info" />
</p>

## Preview

To see all the different possible states of the Webex Avatar component, you can run our Storybook:

```shell
  npm start
```

## Embed

1.  Create a component adapter from which the data will be retrieved (See [adapters](../../adapters)). For instance:

    ```js
    const jsonAdapter = new WebexJSONAdapter(jsonData);
    ```

2.  Create a component instance by passing the person ID as a string and
    enclose it within [a data provider](../WebexDataProvider/WebexDataProvider.js)
    that takes the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously

    ```js
    <WebexDataProvider adapter={jsonAdapter}>
      <WebexMeetingInfo meetingID="meetingID" />
    </WebexDataProvider>
    ```

The component knows how to manage its data. If anything changes in the data source that the adapter manages, the component will also update on its own.
