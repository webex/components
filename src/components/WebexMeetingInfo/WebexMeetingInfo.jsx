import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {format} from 'date-fns';
import {Spinner} from '@momentum-ui/react';

import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import {useMeeting} from '../hooks';

/**
 * Formats a start and end date to a readable string.
 * Example: "5:00 PM - 6:00 PM"
 *
 * @param {Date} startDate  Meeting start time
 * @param {Date} endDate  Meeting end time
 * @returns {string} Formatted time range
 */
export function formatMeetingTime(startDate, endDate) {
  const formattedStartTime = format(startDate, 'p');
  const formattedEndTime = format(endDate, 'p');

  return `${formattedStartTime} - ${formattedEndTime}`;
}

/**
 * Webex Meeting Info component displays the information associated with
 * a given meetingID.
 *
 * @param {object} props  Data passed to the component
 * @param {string} props.meetingID  ID of the meeting
 * @param {string} props.className  Custome CSS class to apply
 * @returns {object} JSX of the component
 */
export default function WebexMeetingInfo({className, meetingID}) {
  const {
    ID,
    startTime,
    endTime,
    title,
  } = useMeeting(meetingID);
  const mainClasses = {
    [`${WEBEX_COMPONENTS_CLASS_PREFIX}-meeting-info`]: true,
    [className]: !!className,
  };
  let infoComponent = <Spinner />;

  if (ID) {
    const displayTitle = title || 'No Meeting Information';

    infoComponent = (
      <>
        <h2>{displayTitle}</h2>
        {
          startTime
          && endTime
          && <h3>{formatMeetingTime(new Date(startTime), new Date(endTime))}</h3>
        }
      </>
    );
  }

  return <div className={classNames(mainClasses)}>{infoComponent}</div>;
}

WebexMeetingInfo.propTypes = {
  className: PropTypes.string,
  meetingID: PropTypes.string.isRequired,
};

WebexMeetingInfo.defaultProps = {
  className: '',
};
