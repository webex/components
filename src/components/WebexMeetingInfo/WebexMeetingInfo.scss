.meeting-info {
  display: flex;
  height: 100%;
  width: 100%;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  text-align: center;
}

.meeting-info h2 {
  font-size: 28px;
  font-family: $brand-font-extra-light;
  color: $md-gray-100;
}

.meeting-info h3 {
  font-family: $brand-font-extra-light;
}
