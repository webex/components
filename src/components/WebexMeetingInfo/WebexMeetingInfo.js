import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {format} from 'date-fns';
import {Spinner} from '@momentum-ui/react';

import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import {useMeeting} from '../hooks';

import './WebexMeetingInfo.scss';

/**
 * Formats a start and end date to a readable string.
 * Example: "5:00 PM - 6:00 PM"
 *
 * @export
 * @param {Date} startDate
 * @param {Date} endDate
 * @returns {String} formatted time range
 */
export function formatMeetingTime(startDate, endDate) {
  const formattedStartTime = format(startDate, 'p');
  const formattedEndTime = format(endDate, 'p');

  return `${formattedStartTime} - ${formattedEndTime}`;
}

/**
 * Webex Meeting Info component displays the information associated with
 * a given meetingID.
 *
 * @param {object} props
 * @returns {object} JSX of the component
 */
export default function WebexMeetingInfo({className, meetingID}) {
  const {ID, startTime, endTime, title} = useMeeting(meetingID);
  let infoComponent;
  const mainClasses = {
    [`${WEBEX_COMPONENTS_CLASS_PREFIX}-meeting-info`]: true,
    [className]: !!className,
  };

  if (ID) {
    const meetingTime = startTime ? <h3>{formatMeetingTime(new Date(startTime), new Date(endTime))}</h3> : null;
    const displayTitle = title || 'No Meeting Information';

    infoComponent = (
      <React.Fragment>
        <h2>{displayTitle}</h2>
        {meetingTime}
      </React.Fragment>
    );
  } else {
    infoComponent = (
      <React.Fragment>
        <Spinner />
      </React.Fragment>
    );
  }

  return <div className={classNames(mainClasses)}>{infoComponent}</div>;
}

WebexMeetingInfo.propTypes = {
  className: PropTypes.string,
  meetingID: PropTypes.string.isRequired,
};

WebexMeetingInfo.defaultProps = {
  className: '',
};
