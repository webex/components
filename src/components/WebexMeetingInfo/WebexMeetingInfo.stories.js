import React from 'react';
import WebexMeetingInfo from './WebexMeetingInfo';

export default {
  title: 'Meetings/Webex Meeting Info',
  component: WebexMeetingInfo,
};

const Template = (args) => <WebexMeetingInfo {...args} />;

export const Loading = Template.bind({});
Loading.args = {
  meetingID: '',
};

export const Error = Template.bind({});
Error.args = {
  meetingID: 'noMeetingTitle',
};

export const Direct = Template.bind({});
Direct.args = {
  meetingID: 'oneOnOneMeeting',
};

export const Space = Template.bind({});
Space.args = {
  meetingID: 'spaceMeeting',
};
