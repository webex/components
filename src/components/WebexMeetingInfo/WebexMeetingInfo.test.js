import React from 'react';

import WebexMeetingInfo, {formatMeetingTime} from './WebexMeetingInfo';

jest.mock('../hooks/useMeeting');

describe('Webex Meeting Info component', () => {
  describe('snapshot', () => {
    test('matches snapshot of scheduled meeting', () => {
      expect(shallow(<WebexMeetingInfo meetingID="scheduledMeeting" />)).toMatchSnapshot();
    });

    test('matches snapshot of scheduled meeting with custom CSS class', () => {
      expect(shallow(<WebexMeetingInfo className="my-custom-class" meetingID="scheduledMeeting" />)).toMatchSnapshot();
    });

    test('matches snapshot of one on one meeting', () => {
      expect(shallow(<WebexMeetingInfo meetingID="oneOnOneMeeting" />)).toMatchSnapshot();
    });

    test('matches snapshot of space meeting', () => {
      expect(shallow(<WebexMeetingInfo meetingID="spaceMeeting" />)).toMatchSnapshot();
    });

    test('matches snapshot of no meeting title', () => {
      expect(shallow(<WebexMeetingInfo meetingID="noMeetingTitle" />)).toMatchSnapshot();
    });

    test('matches snapshot of loading', () => {
      expect(shallow(<WebexMeetingInfo meetingID="loadingMeeting" />)).toMatchSnapshot();
    });
  });

  describe('formatMeetingTime() returns', () => {
    test('a formatted start and end time string', () => {
      const startTime = new Date('2019-08-20T21:00:00.000Z');
      const endTime = new Date('2019-08-20T22:00:00.000Z');

      expect(formatMeetingTime(startTime, endTime)).toEqual('9:00 PM - 10:00 PM');
    });
  });
});
