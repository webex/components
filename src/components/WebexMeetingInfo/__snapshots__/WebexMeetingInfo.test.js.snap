// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webex Meeting Info component snapshot matches snapshot of loading 1`] = `
<div
  className="wxc-meeting-info"
>
  <Spinner
    className=""
    color="black"
    percentage={null}
    showCheck={false}
    showPercentage={false}
    size={36}
  />
</div>
`;

exports[`Webex Meeting Info component snapshot matches snapshot of no meeting title 1`] = `
<div
  className="wxc-meeting-info"
>
  <h2>
    No Meeting Information
  </h2>
</div>
`;

exports[`Webex Meeting Info component snapshot matches snapshot of one on one meeting 1`] = `
<div
  className="wxc-meeting-info"
>
  <h2>
    Jackie Tester's Personal Room
  </h2>
</div>
`;

exports[`Webex Meeting Info component snapshot matches snapshot of scheduled meeting 1`] = `
<div
  className="wxc-meeting-info"
>
  <h2>
    Our Scheduled Meeting
  </h2>
</div>
`;

exports[`Webex Meeting Info component snapshot matches snapshot of scheduled meeting with custom CSS class 1`] = `
<div
  className="wxc-meeting-info my-custom-class"
>
  <h2>
    Our Scheduled Meeting
  </h2>
</div>
`;

exports[`Webex Meeting Info component snapshot matches snapshot of space meeting 1`] = `
<div
  className="wxc-meeting-info"
>
  <h2>
    Ad Hoc Meeting Space
  </h2>
</div>
`;
